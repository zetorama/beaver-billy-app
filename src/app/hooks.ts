import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import type { RootState, AppDispatch } from './store'
import { useEffect, useState } from 'react'

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

/**
 * Check/update if CSS Media Query matches right now
 */
export const useMediaQuery = (query: string) => {
  // Get the initial value of the media query
  const [match, setMatch] = useState(() => window.matchMedia(query).matches)

  // Add a listener for changes and update the state accordingly
  useEffect(() => {
    const mediaQuery = window.matchMedia(query)
    const handleChange = (event: MediaQueryListEvent) => setMatch(event.matches)
    mediaQuery.addEventListener('change', handleChange)
    return () => mediaQuery.removeEventListener('change', handleChange)
  }, [query])

  return match
}

export const useIsMobile = () => !useMediaQuery('(min-width: 576px)')
