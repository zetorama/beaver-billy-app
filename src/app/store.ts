import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit'
import projectsReducer from '../features/projects/Projects.state'

const STORAGE_KEY = 'beaver-app-state'

const getStoredState = () => {
  try {
    const raw = localStorage.getItem(STORAGE_KEY)?.trim()
    return (raw && JSON.parse(raw)) || undefined
  } catch (err) {
    console.warn('Error getting stored state', err)
    return undefined
  }
}

export const store = configureStore({
  reducer: {
    projects: projectsReducer,
  },
  preloadedState: getStoredState(),
})

store.subscribe(() => {
  // Keep whole store in local storage
  const state = store.getState()
  const serializedState = JSON.stringify(state)
  localStorage.setItem(STORAGE_KEY, serializedState)
})

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>
