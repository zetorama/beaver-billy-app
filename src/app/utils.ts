/**
 * Glue two or more strings together with space, keeping only non-empty ones.
 */
export const glue = (...args: unknown[]) => args.filter(Boolean).join(' ')

/**
 * Generate a random string.
 */
export const uuid = () => Math.random().toString(36).slice(2, 12)
