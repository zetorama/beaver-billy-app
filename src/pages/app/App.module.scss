.root {
  min-height: 100vh;
  position: relative;
}

.header {
  color: #424242;
  background-color: #f7f9fd;
  box-shadow: 0px -1px 4px rgba(0, 0, 0, 0.203436);
}

.headerOuter {
  position: relative;
  margin: auto;

  @media (min-width: 576px) {
    max-width: 85%;
  }

  @media (min-width: 1600px) {
    max-width: (1600px * 0.85);
  }
}

.headerInner {
  padding: 12px 16px;
}

.main {
  margin: 60px auto 120px;
  background-color: #fdfdfd;
  color: #4c4c4c;

  @media (min-width: 576px) {
    margin-bottom: 60px;
    max-width: 85%;
  }

  @media (min-width: 1600px) {
    max-width: (1600px * 0.85);
  }
}

.logo {
  display: block;
  width: 2rem;
  height: 2rem;

  background-image: url(./logo.png);
  background-repeat: no-repeat;
  background-size: cover;
}

.headline {
  margin: 0;
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;

  .logo + & {
    margin-top: 48px;
  }
}

.newProjectButton {
  width: 60px !important;
  height: 60px !important;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #4a475f;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);

  &.isOnEdge {
    position: absolute;
    bottom: 0;
    right: 0;
    transform: translateY(50%);
  }

  &.isInCorner {
    position: fixed;
    bottom: 40px;
    right: 16px;
  }
}
