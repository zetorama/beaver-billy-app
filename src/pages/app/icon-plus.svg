<svg xmlns="http://www.w3.org/2000/svg" width="18" height="18" fill="none">
  <rect width="1.636" height="18" x="18" y="8.182" fill="#fff" rx=".818" transform="rotate(90 18 8.182)"/>
  <rect width="1.636" height="18" x="9.818" y="18" fill="#fff" rx=".818" transform="rotate(180 9.818 18)"/>
</svg>