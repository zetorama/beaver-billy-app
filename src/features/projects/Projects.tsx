import { List } from 'antd'
import { useAppDispatch, useAppSelector, useIsMobile } from '~/app/hooks'
import ProjectCard, { NewProjectCard } from './ProjectCard'
import { selectAllProjects, selectIsNewProjectRequested, toggleNewProjectRequest } from './Projects.state'
import styles from './Projects.module.scss'

export default function Projects() {
  const isMobile = useIsMobile()
  const projects = useAppSelector(selectAllProjects)
  const isAddingNewProject = useAppSelector(selectIsNewProjectRequested)

  if (!projects.length && !isAddingNewProject) {
    return <ProjectsPlaceholder />
  }

  // TODO: figure out reordering (drag'n'drop) and pagination (infinite scroll?)

  return (
    <List
      bordered
      className={styles.root}
      dataSource={projects}
      renderItem={(project) => (
        <List.Item key={project.id} className={styles.row}>
          <ProjectCard project={project} isCompact={isMobile} />
        </List.Item>
      )}
      header={
        isAddingNewProject && (
          <List.Item className={styles.row}>
            <NewProjectCard isCompact={isMobile} />
          </List.Item>
        )
      }
      locale={{ emptyText: 'No projects yet' }}
    />
  )
}

function ProjectsPlaceholder() {
  const dispatch = useAppDispatch()
  const handleClick = () => {
    dispatch(toggleNewProjectRequest())
  }

  return (
    <div className={styles.placeholder}>
      <h1>No projects yet</h1>
      <p>
        Click{' '}
        <button onClick={handleClick} aria-label='Create new project'>
          âž•
        </button>{' '}
        to create your first project.
      </p>
    </div>
  )
}
