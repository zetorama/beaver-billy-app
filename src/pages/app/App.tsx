import { Button, Layout } from 'antd'
import { useAppDispatch, useIsMobile } from '~/app/hooks'
import { toggleNewProjectRequest } from '~/features/projects/Projects.state'
import Projects from '~/features/projects/Projects'
import { ReactComponent as IconPlus } from './icon-plus.svg'
import styles from './App.module.scss'
import { glue } from '~/app/utils'

export default function App() {
  const isMobile = useIsMobile()

  return (
    <Layout className={styles.root}>
      <header className={styles.header}>
        <div className={styles.headerOuter}>
          <div className={styles.headerInner}>
            <div className={styles.logo}>
              <span className='sr-only'>Beaver Billy's App</span>
            </div>
            <h1 className={styles.headline}>My projects</h1>

            {!isMobile && <NewProjectButton className={styles.isOnEdge} />}
          </div>
        </div>
      </header>
      <Layout.Content>
        <main className={styles.main}>
          <Projects />
        </main>

        {isMobile && <NewProjectButton className={styles.isInCorner} />}
      </Layout.Content>
    </Layout>
  )
}

function NewProjectButton({ className }: { className?: string }) {
  const dispatch = useAppDispatch()
  const handleClick = () => {
    dispatch(toggleNewProjectRequest())
    window.scrollTo({ top: 0, behavior: 'smooth' })
  }

  return (
    <Button
      type='primary'
      shape='circle'
      size='large'
      aria-label='Create new project'
      onClick={handleClick}
      icon={<IconPlus />}
      rootClassName={glue(styles.newProjectButton, className)}
    />
  )
}
