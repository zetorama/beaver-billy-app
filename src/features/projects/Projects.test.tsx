import { fireEvent, render, screen, waitForElementToBeRemoved } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import { reducer, store as emptyStore } from '~/app/store'
import Projects from './Projects'

// eslint-disable-next-line jest/no-mocks-import
import mockedProjects from './__mocks__/projects-list.json'

const formatDate = (date: string) =>
  new Date(date === 'now' ? Date.now() : date).toLocaleDateString(undefined, {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  })

const store = configureStore({
  reducer,
  preloadedState: {
    projects: { items: mockedProjects },
  },
})

describe('Projects', () => {
  it('renders all my projects', () => {
    const { container } = render(
      <Provider store={store}>
        <Projects />
      </Provider>,
    )

    expect(container).toMatchSnapshot()
  })

  it('updates project name', async () => {
    const project = mockedProjects[0]

    render(
      <Provider store={store}>
        <Projects />
      </Provider>,
    )

    // get an edit project form
    const button = screen.getAllByRole('button', { name: /rename project/i })[0]
    expect(button).toBeInTheDocument()
    fireEvent.click(button)

    // type in project name
    const input = screen.getByRole('textbox', { name: /project name/i })
    expect(input).toHaveValue(project.title)
    fireEvent.change(input, { target: { value: 'My Fancy Project' } })

    // submit form
    fireEvent.click(screen.getByRole('button', { name: /submit/i }))

    await screen.findByText(/my fancy project/i)

    // check there's a new project's name and not previous name
    expect(screen.getByText(/my fancy project/i)).toBeInTheDocument()
    expect(screen.queryByText(project.title)).not.toBeInTheDocument()
  })

  it('deletes project', async () => {
    const lastIndex = mockedProjects.length - 1
    const project = mockedProjects[lastIndex]

    render(
      <Provider store={store}>
        <Projects />
      </Provider>,
    )

    expect(screen.getAllByRole('button', { name: /delete project/i })).toHaveLength(mockedProjects.length)

    // click on delete button
    const button = screen.getAllByRole('button', { name: /delete project/i })[lastIndex]
    expect(button).toBeInTheDocument()
    fireEvent.click(button)

    // confirm deletion
    await screen.findByText(/are you sure/i)
    fireEvent.click(screen.getByRole('button', { name: /yes/i }))

    // check it's gone
    await waitForElementToBeRemoved(() => screen.queryByText(project.title))

    expect(screen.queryByText(project.title)).not.toBeInTheDocument()
    expect(screen.getAllByRole('button', { name: /delete project/i })).toHaveLength(mockedProjects.length - 1)
  })

  it('creates a new project', async () => {
    render(
      <Provider store={emptyStore}>
        <Projects />
      </Provider>,
    )

    // get new project form
    const button = screen.getByRole('button', { name: /new project/i })
    expect(button).toBeInTheDocument()
    fireEvent.click(button)

    // type in project name
    const input = screen.getByPlaceholderText(/project name/i)
    expect(input).toBeInTheDocument()
    fireEvent.change(input, { target: { value: 'My Fancy Project' } })

    // submit form
    fireEvent.click(screen.getByRole('button', { name: /submit/i }))

    await screen.findByText(/my fancy project/i)

    // check project's name and creation date
    expect(screen.getByText(/my fancy project/i)).toBeInTheDocument()
    expect(screen.getByText(formatDate('now'))).toBeInTheDocument()
  })
})
