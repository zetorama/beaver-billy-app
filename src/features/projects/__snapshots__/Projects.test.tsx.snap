// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Projects > renders all my projects 1`] = `
<div>
  <div
    class="ant-list ant-list-split ant-list-bordered _root_877127 css-dev-only-do-not-override-ed5zg0"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-ed5zg0"
    >
      <div
        class="ant-spin-container"
      >
        <ul
          class="ant-list-items"
        >
          <li
            class="ant-list-item _row_877127"
          >
            <div
              class="_root_85982d _isCompact_85982d"
            >
              <div
                class="_picture_85982d"
              >
                <img
                  alt=""
                  class="_pictureImg_85982d"
                  src="/features/projects/project-default-picture.png"
                />
              </div>
              <div
                class="_description_85982d"
              >
                <h2
                  class="_descriptionTitle_85982d"
                >
                  Dam Good
                </h2>
                <div
                  class="_descriptionTimestamp_85982d"
                >
                  <span
                    style="margin-right: 1.5ch;"
                  >
                    Feb 11, 2023
                  </span>
                  <span>
                    5:12 p.m.
                  </span>
                </div>
              </div>
              <div
                class="_actions_85982d"
              >
                <button
                  aria-label="Rename project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20.0248 19.3244H3.97419C3.58656 19.3244 3.27295 19.638 3.27295 20.0257C3.27295 20.4134 3.58656 20.727 3.97419 20.727H20.0248C20.4124 20.727 20.726 20.4134 20.726 20.0257C20.726 19.638 20.4124 19.3244 20.0248 19.3244ZM7.79205 17.0256L12.3793 15.0833C12.4533 15.0522 12.5196 15.0074 12.578 14.9509L20.315 7.213C20.8624 6.66558 20.8663 5.77335 20.3189 5.22593L18.7742 3.68108C18.2269 3.13366 17.3348 3.13756 16.7874 3.68498L9.05039 11.4229C8.9939 11.4793 8.9491 11.5475 8.91793 11.6216L6.97394 16.2074C6.86485 16.4646 6.94082 16.7236 7.10639 16.8912C7.27196 17.0587 7.53298 17.1347 7.79205 17.0256ZM17.7789 4.67656L17.7828 4.67267L19.3255 6.21557L19.3216 6.21946L18.1139 7.42729L16.5712 5.88439L17.7789 4.67656ZM10.149 12.3073L15.5797 6.87597L17.1224 8.41887L11.6917 13.8502L9.01532 14.984L10.149 12.3073Z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Delete project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.46154 5.25V4.125C7.46154 3.50368 7.97813 3 8.61538 3H14.3846C15.0219 3 15.5385 3.50368 15.5385 4.125V5.25H19V6.375H4V5.25H7.46154ZM5.15385 7.5H17.8462V19.875C17.8462 20.4963 17.3296 21 16.6923 21H6.30769C5.67044 21 5.15385 20.4963 5.15385 19.875V7.5ZM7.46154 9.75V18.75H8.61538V9.75H7.46154ZM10.9231 9.75V18.75H12.0769V9.75H10.9231ZM14.3846 9.75V18.75H15.5385V9.75H14.3846Z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </li>
          <li
            class="ant-list-item _row_877127"
          >
            <div
              class="_root_85982d _isCompact_85982d"
            >
              <div
                class="_picture_85982d"
              >
                <img
                  alt=""
                  class="_pictureImg_85982d"
                  src="/features/projects/project-default-picture.png"
                />
              </div>
              <div
                class="_description_85982d"
              >
                <h2
                  class="_descriptionTitle_85982d"
                >
                  Chewy & Woody
                </h2>
                <div
                  class="_descriptionTimestamp_85982d"
                >
                  <span
                    style="margin-right: 1.5ch;"
                  >
                    Mar 12, 2023
                  </span>
                  <span>
                    6:13 p.m.
                  </span>
                </div>
              </div>
              <div
                class="_actions_85982d"
              >
                <button
                  aria-label="Rename project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20.0248 19.3244H3.97419C3.58656 19.3244 3.27295 19.638 3.27295 20.0257C3.27295 20.4134 3.58656 20.727 3.97419 20.727H20.0248C20.4124 20.727 20.726 20.4134 20.726 20.0257C20.726 19.638 20.4124 19.3244 20.0248 19.3244ZM7.79205 17.0256L12.3793 15.0833C12.4533 15.0522 12.5196 15.0074 12.578 14.9509L20.315 7.213C20.8624 6.66558 20.8663 5.77335 20.3189 5.22593L18.7742 3.68108C18.2269 3.13366 17.3348 3.13756 16.7874 3.68498L9.05039 11.4229C8.9939 11.4793 8.9491 11.5475 8.91793 11.6216L6.97394 16.2074C6.86485 16.4646 6.94082 16.7236 7.10639 16.8912C7.27196 17.0587 7.53298 17.1347 7.79205 17.0256ZM17.7789 4.67656L17.7828 4.67267L19.3255 6.21557L19.3216 6.21946L18.1139 7.42729L16.5712 5.88439L17.7789 4.67656ZM10.149 12.3073L15.5797 6.87597L17.1224 8.41887L11.6917 13.8502L9.01532 14.984L10.149 12.3073Z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Delete project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.46154 5.25V4.125C7.46154 3.50368 7.97813 3 8.61538 3H14.3846C15.0219 3 15.5385 3.50368 15.5385 4.125V5.25H19V6.375H4V5.25H7.46154ZM5.15385 7.5H17.8462V19.875C17.8462 20.4963 17.3296 21 16.6923 21H6.30769C5.67044 21 5.15385 20.4963 5.15385 19.875V7.5ZM7.46154 9.75V18.75H8.61538V9.75H7.46154ZM10.9231 9.75V18.75H12.0769V9.75H10.9231ZM14.3846 9.75V18.75H15.5385V9.75H14.3846Z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </li>
          <li
            class="ant-list-item _row_877127"
          >
            <div
              class="_root_85982d _isCompact_85982d"
            >
              <div
                class="_picture_85982d"
              >
                <img
                  alt=""
                  class="_pictureImg_85982d"
                  src="/features/projects/project-default-picture.png"
                />
              </div>
              <div
                class="_description_85982d"
              >
                <h2
                  class="_descriptionTitle_85982d"
                >
                  BeaverBookClubForFurryFriends
                </h2>
                <div
                  class="_descriptionTimestamp_85982d"
                >
                  <span
                    style="margin-right: 1.5ch;"
                  >
                    Apr 13, 2023
                  </span>
                  <span>
                    6:14 p.m.
                  </span>
                </div>
              </div>
              <div
                class="_actions_85982d"
              >
                <button
                  aria-label="Rename project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20.0248 19.3244H3.97419C3.58656 19.3244 3.27295 19.638 3.27295 20.0257C3.27295 20.4134 3.58656 20.727 3.97419 20.727H20.0248C20.4124 20.727 20.726 20.4134 20.726 20.0257C20.726 19.638 20.4124 19.3244 20.0248 19.3244ZM7.79205 17.0256L12.3793 15.0833C12.4533 15.0522 12.5196 15.0074 12.578 14.9509L20.315 7.213C20.8624 6.66558 20.8663 5.77335 20.3189 5.22593L18.7742 3.68108C18.2269 3.13366 17.3348 3.13756 16.7874 3.68498L9.05039 11.4229C8.9939 11.4793 8.9491 11.5475 8.91793 11.6216L6.97394 16.2074C6.86485 16.4646 6.94082 16.7236 7.10639 16.8912C7.27196 17.0587 7.53298 17.1347 7.79205 17.0256ZM17.7789 4.67656L17.7828 4.67267L19.3255 6.21557L19.3216 6.21946L18.1139 7.42729L16.5712 5.88439L17.7789 4.67656ZM10.149 12.3073L15.5797 6.87597L17.1224 8.41887L11.6917 13.8502L9.01532 14.984L10.149 12.3073Z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Delete project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.46154 5.25V4.125C7.46154 3.50368 7.97813 3 8.61538 3H14.3846C15.0219 3 15.5385 3.50368 15.5385 4.125V5.25H19V6.375H4V5.25H7.46154ZM5.15385 7.5H17.8462V19.875C17.8462 20.4963 17.3296 21 16.6923 21H6.30769C5.67044 21 5.15385 20.4963 5.15385 19.875V7.5ZM7.46154 9.75V18.75H8.61538V9.75H7.46154ZM10.9231 9.75V18.75H12.0769V9.75H10.9231ZM14.3846 9.75V18.75H15.5385V9.75H14.3846Z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </li>
          <li
            class="ant-list-item _row_877127"
          >
            <div
              class="_root_85982d _isCompact_85982d"
            >
              <div
                class="_picture_85982d"
              >
                <img
                  alt=""
                  class="_pictureImg_85982d"
                  src="/features/projects/project-default-picture.png"
                />
              </div>
              <div
                class="_description_85982d"
              >
                <h2
                  class="_descriptionTitle_85982d"
                >
                  Tales about tails
                </h2>
                <div
                  class="_descriptionTimestamp_85982d"
                >
                  <span
                    style="margin-right: 1.5ch;"
                  >
                    May 14, 2023
                  </span>
                  <span>
                    6:15 p.m.
                  </span>
                </div>
              </div>
              <div
                class="_actions_85982d"
              >
                <button
                  aria-label="Rename project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20.0248 19.3244H3.97419C3.58656 19.3244 3.27295 19.638 3.27295 20.0257C3.27295 20.4134 3.58656 20.727 3.97419 20.727H20.0248C20.4124 20.727 20.726 20.4134 20.726 20.0257C20.726 19.638 20.4124 19.3244 20.0248 19.3244ZM7.79205 17.0256L12.3793 15.0833C12.4533 15.0522 12.5196 15.0074 12.578 14.9509L20.315 7.213C20.8624 6.66558 20.8663 5.77335 20.3189 5.22593L18.7742 3.68108C18.2269 3.13366 17.3348 3.13756 16.7874 3.68498L9.05039 11.4229C8.9939 11.4793 8.9491 11.5475 8.91793 11.6216L6.97394 16.2074C6.86485 16.4646 6.94082 16.7236 7.10639 16.8912C7.27196 17.0587 7.53298 17.1347 7.79205 17.0256ZM17.7789 4.67656L17.7828 4.67267L19.3255 6.21557L19.3216 6.21946L18.1139 7.42729L16.5712 5.88439L17.7789 4.67656ZM10.149 12.3073L15.5797 6.87597L17.1224 8.41887L11.6917 13.8502L9.01532 14.984L10.149 12.3073Z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Delete project"
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-circle ant-btn-link ant-btn-icon-only"
                  type="button"
                >
                  <svg
                    class="_icon_85982d"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.46154 5.25V4.125C7.46154 3.50368 7.97813 3 8.61538 3H14.3846C15.0219 3 15.5385 3.50368 15.5385 4.125V5.25H19V6.375H4V5.25H7.46154ZM5.15385 7.5H17.8462V19.875C17.8462 20.4963 17.3296 21 16.6923 21H6.30769C5.67044 21 5.15385 20.4963 5.15385 19.875V7.5ZM7.46154 9.75V18.75H8.61538V9.75H7.46154ZM10.9231 9.75V18.75H12.0769V9.75H10.9231ZM14.3846 9.75V18.75H15.5385V9.75H14.3846Z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
